### DISCUSSION TITLE
Add rotation to signature content

### DISCUSSION THREAD

**Discussion by crodbor** [MatthiasValvekens/pyHanko#202] [Ideas]
I've had to add a signature field to some PDF in landscape orientation, so I used the standard approach with the `append_signature_field` function. It worked but for some reason the origin of coordinates was at the top left, which is not a serious problem.
However, after the field was signed the text inside was rotated. After some test with other PDF in the same orientation I arrived to the conclusion that the ones I wanted to add the field behaved as in portrait orientation and then rotated, which I assume is due to the program that generates them.
In order to let the signature be afterwards in the correct orientation I had to patch the `__init__` method of the `SignatureFormField`class like this:
```
from functools import wraps
from pyhanko.pdf_utils import generic
from pyhanko.sign.fields import SignatureFormField

def rotate_field(rotation, func):
    @wraps(func)
    def wrapper(cls, *args, **kwargs):
        func(cls, *args, **kwargs)
        # Parameter of the Subtype 'Widget' to add afterwards the rotation of the content
        cls.annot_dict['/MK'] = custom_dict = generic.DictionaryObject()
        custom_dict['/R'] = generic.NumberObject(rotation)
    return wrapper

SignatureFormField.__init__ = rotate_field(90, SignatureFormField.__init__)
```
The problem with this approach is that the behavior is still fixed and not customizable for different cases.

It would be interesting to add the option of rotation to the API, maybe in the `VisibleSigSettings` class.

**Comment by MatthiasValvekens** [comment]
Yeah, that's because the interaction between page rotation settings and annotations is a bit wonky in PDF. It's notoriously hard to write code that magically does the right thing in all cases (side note: I personally believe that adding a separate page rotation control to the file format _outside_ of the standard graphics model was a design mistake in hindsight, but that ship has long sailed).

IIRC Acrobat even just compensates for page rotation in the signature appearance stream directly (by prepending it with the inverse rotation transformation). pyHanko could perhaps do the same thing by default, but it's something for which I would need to find time to investigate. I'd welcome a PR adding such functionality, though.

Just to make sure: did you try experimenting with the `NoRotate` flag?

**Reply by crodbor** [reply]
I tried using the `NoRotate` flag, but turning it to `False` made the signature field disappear, so it wasn't the solution. I also tried all the other flags available during creation, but I suppose it was due to the weird behavior of the PDF file.

In the end I rotated the content of the signature, not the annotation itself, this way, you can tweak manually the end result in case of unexpected behavior. Also, it will allow to rotate the content to show it different in expected behavior.

As you mention, doing it by default won't be easy, so my proposition is just a configurable feature of the visible signature, which afterwards in can be use in those cases.

I'd like to write a PR, but I only have experience with scripting for my personal use, not libraries or projects this big. In fact, this is my first post in github. I just found where to add the patch for my case and I saw it as an useful feature for everyone, however, I don't know the library enough to propagate upwards where the flag/option should be added or the validity checks for the values, which in the end is the trickiest part.

**Reply by MatthiasValvekens** [reply]
Makes sense. I can't promise when I'll have time to get to the bottom of this, but feel free to file an `enhancement` issue on the issue tracker for a feature to rotate signature contents. Helps us to not lose track :).

**Comment by eduperottoni** [comment]
Hi, @crodbor and @MatthiasValvekens 

I have the same problem as @crodbor and I identified the same behavior in some PDF in landscape orientation. For now, while pyHanko doesn't support it, I was trying to rotate the content, doing what @crodbor have shown above. I also use `append_signature_field` function so I thought the @crodbor's code would solve the problem, but it didn't.

To make the test, I rotated a PDF 90 degrees clockwise in `www.ilovepdf.com` and made a signature, with `rotation` parameter of `rotate_field` setted as 90 (to correct the rotation of the page, I have to rotate the field 90 degrees counterclockwise). The signed document is the following:

[signed_document.pdf](https://github.com/MatthiasValvekens/pyHanko/files/12111604/signed_document.pdf)

After signing the document, the `/Annot` dictionary contains the `/MK` dict and the `/MK` dict contains the correct `/R` value, wich is done by `rotate_field` function. Despite that, opening the file in PDF viewers, visually, nothing has changed. 

Is the `/MK` dictionary modification enough to ensure the rotation of the stamp's content? Besides the @crodbor's code, is there anything that should be done? I'd appreciate your help...

Thanks a lot.

**Reply by MatthiasValvekens** [reply]
Hi @eduperottoni 

Can't really run any tests myself since I'm currently in bed recovering from a surprisingly persistent infection, but have you compared your strategy with the proposed changes in #266?

I would personally deal with this at the content stream level instead (by prepending a rotation transformation to the signature appearance stream), but if you can solve it using "higher level" settings, then that works for me as well.

**Reply by eduperottoni** [reply]
Hi, @MatthiasValvekens 

First of all, I wish you a good and speedy recovery. Health in 1st place, always.

No, I have not compared the strategies you mentioned. I just followed the @crodbor's code, which didn't work.

As soon as I can, I will try to work in the content stream level and I'd appreciate if you could take a look on that, since it doesn't seem to be simple by reading the comments in #266.

**Comment by maxkurze1** [comment]
Hey guys;

I basically created a mixed cli/gui frontend for pyHanko that comes with a couple of usability features. These features include rotation recognition and automatic adjustment of the signature.

This frontend just rotates the whole content stream of the signature in case its page contains a rotation. 


Unfortunately it also has a lot of other usability bloat - but if dou don't mind that, you should check it out:

https://github.com/maxkurze1/resignation

PS: It is still a rather early version, but I would be very happy about some feedback.
